ou_dedetai/tui_app.py

```diff
@@ -719,25 +713,25 @@ def switch_screen(self, dialog):
         if isinstance(self.active_screen, tui_screen.CursesScreen):
             self.clear()
 
-    def get_product(self, dialog):
-        question = "Choose which FaithLife product the script should install:"  # noqa: E501
-        labels = ["Logos", "Verbum", "Return to Main Menu"]
-        options = self.which_dialog_options(labels, dialog)
+    _exit_option = "Return to Main Menu"
+
+    def _ask(self, question: str, options: list[str]) -> Optional[str]:
+        options = self.which_dialog_options(options, config.use_python_dialog)
         self.menu_options = options
-        self.screen_q.put(self.stack_menu(2, self.product_q, self.product_e, question, options, dialog=dialog))
+        self.screen_q.put(self.stack_menu(2, Queue(), threading.Event(), question, options, dialog=config.use_python_dialog))
 
-    def set_product(self, choice):
-        if str(choice).startswith("Logos"):
-            config.FLPRODUCT = "Logos"
-        elif str(choice).startswith("Verbum"):
-            config.FLPRODUCT = "Verbum"
-        self.menu_screen.choice = "Processing"
-        self.product_q.put(config.FLPRODUCT)
-        self.product_e.set()
+        # Now wait for it to complete
+        self.ask_answer_event.wait()
+        return self.ask_answer_queue.get()
+
+    def handle_ask_response(self, choice: Optional[str]):
+        if choice is not None:
+            self.ask_answer_queue.put(choice)
+            self.ask_answer_event.set()
+            self.switch_screen(config.use_python_dialog)
 
     def get_version(self, dialog):
-        self.product_e.wait()
-        question = f"Which version of {config.FLPRODUCT} should the script install?"  # noqa: E501
+        question = f"Which version of {self.conf.faithlife_product} should the script install?"  # noqa: E501
```

Comment:
To my knowledge, I need to wait on self.product_e.wait() as if not, the TUI charges through the installer process; should this way now be handled by the TUI's implementation of app._hook_product_update()?

Reply:
Surprisingly enough it didn't charge through - it waited for the question to be answered before going to the next screen probably because we have two main threads in the TUI:

1. The input processing thread (main, don't block me)
2. When events are triggered they're spawned on a new thread (this is the installation thread). Since the ask call is blocking, this gets stopped while the user is inputting their value (which the processing is done on the first thread)

Then if you notice in TUI's ask implementation there is an event wait to communicate between the two threads